yarn-project/pxe/src/storage/note_data_provider/note_data_provider.test.ts

```diff
@@ -1,469 +1,581 @@
+        status: NoteStatus.ACTIVE_OR_NULLIFIED,
       });
+      expect(bySlot.filter(n => n.index === BigInt(1))).toHaveLength(1);
+
+      const byScope = await provider.getNotes({
+        contractAddress: contractA,
+        scopes: [note1.recipient],
+        status: NoteStatus.ACTIVE_OR_NULLIFIED,
+      });
+      expect(byScope.filter(n => n.index === BigInt(1))).toHaveLength(1);
+    });
+
+    it('attempts to nullify the same note twice in succession results in error', async () => {
+      await provider.applyNullifiers([mkNullifier(note1)]); // First application should succeed
+      const activeNotes = await provider.getNotes({ contractAddress: contractA });
+      expect(activeNotes.map(n => n.index.toString()).sort()).toEqual(['2']);
+
+      // should throw on second attempt as note1 is already nullified
+      await expect(provider.applyNullifiers([mkNullifier(note1)])).rejects.toThrow(
+        'Nullifier not found in applyNullifiers',
+      );
+    });
+
+    it('attempts to nullify the same note twice in same call results in error', async () => {
+      const nullifiers = [mkNullifier(note1), mkNullifier(note1)];
+      await expect(provider.applyNullifiers(nullifiers)).rejects.toThrow('Nullifier not found in applyNullifiers');
     });
   });
 
-  // --- Account Scoping Tests ---
-
-  describe('Account scoping and Account-scoped note storage', () => {
-    /**
-     * Account-scoped note storage and retrieval test.
-     *
-     * Tests that notes can be stored and retrieved per specific account (recipient),
-     * implementing proper privacy isolation between users.
-     * This test:
-     * 1. Stores first 5 notes under recipient[0]'s scope
-     * 2. Stores remaining 5 notes under recipient[1]'s scope
-     * 3. Queries notes scoped to recipient[0] - should get only first 5 notes
-     * 4. Queries notes scoped to recipient[1] - should get only last 5 notes
-     * 5. Queries notes scoped to both recipients - should get all 10 notes
-     */
-    it('stores notes and retrieves notes with siloed account', async () => {
-      await noteDataProvider.addNotes(notes.slice(0, 5), recipients[0]);
-      await noteDataProvider.addNotes(notes.slice(5), recipients[1]);
-
-      const recipient0Notes = await getNotesForAllContracts({
-        scopes: [recipients[0]],
+  describe('NoteDataProvider.rollbackNotesAndNullifiers', () => {
+    let provider: NoteDataProvider;
+    let store: AztecLMDBStoreV2;
+
+    beforeEach(async () => {
+      store = await openTmpStore('note_data_provider_rollback_test');
+      provider = await NoteDataProvider.create(store);
+      await provider.addScope(scope1);
+      await provider.addScope(scope2);
+    });
+
+    afterEach(async () => {
+      await store.close();
+    });
+
+    describe('rewind nullifications happy path', () => {
+      async function setupRollbackScenario() {
+        const noteBlock1 = await mkNote({ index: BigInt(1), l2BlockNumber: 1 }); // Nullified at block 2
```

Comment:
```suggestion
        const noteBlock1 = await mkNote({ index: 1n, l2BlockNumber: 1 }); // Nullified at block 2
```
Just a detail but we generally use this syntax for instantiating a bigint in the repo and I think it's easier to read.